File: C++/380. Insert Delete GetRandom O(1).cpp
class RandomizedSet {
    private:
        vector<int> nums; 
        unordered_map<int,int> x; 
    public:
        RandomizedSet() {
        }
        bool insert(int value) {
            if (x.find(value) != x.end()) {return false;}
            nums.emplace_back(value);
            x[value] = nums.size() -1;
            return true;
        }
        bool remove(int value) {
            if (x.find(value) == x.end()){return false;}
            int last = nums.back();
            x[last] = x[value]; 
            nums[x[value]] = last; 
            nums.pop_back();
            x.erase(value);
            return true;
        }
        int getRandom() {
            return nums[rand() % nums.size()];
        }
};

File: JAVA/380. Insert Delete GetRandom O(1).java
class RandomizedSet {
    private Map<Integer, Integer> map; 
    private List<Integer> list; 
    private Random rand;
    public RandomizedSet() {
        map = new HashMap<>();
        list = new ArrayList<>();
        rand = new Random();
    }
    public boolean insert(int val) {
        if(map.containsKey(val)) {
            return false;
        }
        map.put(val, list.size());
        list.add(val);
        return true;
    }
    public boolean remove(int val) {
        if(!map.containsKey(val)) {
            return false;
        }
        int lastElement = list.get(list.size() - 1);
        int idx = map.get(val);        
        list.set(idx, lastElement);
        map.put(lastElement, idx);        
        list.remove(list.size() - 1);
        map.remove(val);        
        return true;
    }
    public int getRandom() {
        return list.get(rand.nextInt(list.size()));
    }
}
